### PR TITLE
Fixed keepLastValue with quoteless INF

### PR DESCRIPTION
**Pull request by DeleurApps**
https://github.com/grafana/metrictank/issues/1311

### CODE REVIEW

**Comment by DanCech** [comment]
graphite-web doesn't handle a bare `INF` properly either, it interprets it as a series path, tries to find the series `INF`, and passes an empty array to `keepLastValue`.  We should really fix the bug in graphite-web rather than duplicating the bad behavior here.  To do that we'll need to update `webapp/render/grammar.py` to properly interpret `INF`, and update `keepLastValue` to stop interpreting random strings as meaning no limit.  Longer-term I'd like to see `keepLastValue` updated to support specifying a time range string in addition to a number of points.

**Comment by DeleurApps** [comment]
@shanson7 thoughts on this?

I agree that graphite should be fixed, but for now, if we are trying to replicate graphite functionality, we should stick with what currently exists there. Also, there seems to be a different problem where grafana removes quotes from `INF` as mentioned in https://github.com/grafana/metrictank/issues/1311

I can definitely add a validator that checks the quoted/unquoted string so that only `"INF"` or `INF` is accepted. This would be more correct behavior, but wouldn't completely match graphite.

**Comment by shanson7** [comment]
The fact of the matter is that bare INF likely exists in many saved dashboards now due to the Grafana bug, and it effectively works through graphite and doesn't through metrictank. I'm inclined to make it officially supported.

**Comment by Dieterbe** [comment]
> The fact of the matter is that bare INF likely exists in many saved dashboards now due to the Grafana bug

while this sounds true, do you have a link to grafana bug? i couldn't find it. some context would help here. 

regarding what Dan said:

> graphite-web doesn't handle a bare INF properly either, it interprets it as a series path, tries to find the series INF, and passes an empty array to keepLastValue

does this bug effectively (and coincidentally) result in the same behavior as 'INF' would?
I'm asking because Sean said "it effectively works through graphite".

> We should really fix the bug in graphite-web rather than duplicating the bad behavior here

aren't we conflating two things here? Without having dug deep I agree with you that we should make graphite handle INF properly (meaning handle it in addition to 'INF') and i think we should do the same for MT, because as sean said, the dashboard queries likely already exist.

**Comment by DeleurApps** [comment]
> > graphite-web doesn't handle a bare INF properly either, it interprets it as a series path, tries to find the series INF, and passes an empty array to keepLastValue
> 
> does this bug effectively (and coincidentally) result in the same behavior as 'INF' would?
> I'm asking because Sean said "it effectively works through graphite".

yes, it does. Essentially just falls back to the default value (INF) without throwing an error.

**Comment by Dieterbe** [comment]
Note that the syntax described on  https://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.keepLastValue uses an unquoted inf.
i couldn't reproduce what Sean said ("When editing queries in grafana, the quotes are frequently removed, breaking the whole query.")
seems that with the UI based editor it by default puts `INF` without quotes, but the request include quotes, and also if you change edit mode it shows with quotes. even after saving and reloading that remains consistent.

anyway, Graphite documents it without quotes, it works in graphite (due to a bug or not) and we should support it the same.

File: expr/func_keeplastvalue.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ func (s *FuncKeepLastValue) Signature() ([]Arg, []Arg) {
 					// Treats any string as infinity. This matches Graphite's behavior
 					// (although intended bevahior is to let user specify "INF" as the limit)
 					ArgString{val: &stub},
+					ArgQuotelessString{val: &stub},
```
Comment (Dieterbe) [inline expr/func_keeplastvalue.go]: are we declaring here that we need a string argument followed by a quotelessstring argument?
this doesn't seem to make sense. should we use `ArgIn` (which is what you'd use if you want to allow one of more types of args)

**Comment by Dieterbe** [comment]
As a simpler solution we could also rewrite all input queries by replacing `,INF)` with `,`INF`)` with

File: expr/func_keeplastvalue.go (hunk 1 above)
Comment (Dieterbe) [reply expr/func_keeplastvalue.go]: d'oh, i didn't expand the diff enough. this is precisely what you're doing...

**Comment by Dieterbe** [comment]
OK I'm good with this. thanks @DeleurApps !

**Pull request merged**